pkg/cmd/util/clientcmd/factory.go

```diff
@@ -774,3 +776,46 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
 	c.clients[config.GroupVersion.String()] = client
 	return client, nil
 }
+
+// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
+// and are not subresources. This is the closest mapping possible from the client side to resources that can be
+// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
+// represented. It collapses resources in different groups with the same name into a single value.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(d discovery.Discovery, m meta.RESTMapper) ([]unversioned.GroupResource, error) {
+	set := make(map[string]unversioned.GroupResource)
```

Comment:
> I don't think they should (it's not the point of this method).  The
> heuristic is resources with the same name share storage, and are henceforth
> only migrated once.

`roles.v1.legacyopenshift` and `roles.v1beta1.rbac.k8s.io` would seem to be a counterexample.  I see no reason to believe we won't collide more in the future.